pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -231,29 +231,31 @@ private CompletableFuture<Void> addAcknowledgment(MessageIdAdv msgId,
             case Individual:
                 return addIndividualAcknowledgment(msgId,
                         batchMessageId,
-                        __ -> doIndividualAck(__, properties),
-                        __ -> doIndividualBatchAck(__, properties));
+                        __ -> doIndividualAck(__, properties, false),
+                        __ -> doIndividualBatchAck(__, properties, false));
             case Cumulative:
                 if (batchMessageId != null) {
                     consumer.onAcknowledgeCumulative(batchMessageId, null);
                 } else {
                     consumer.onAcknowledgeCumulative(msgId, null);
                 }
                 if (batchMessageId == null || MessageIdAdvUtils.acknowledge(batchMessageId, false)) {
-                    return doCumulativeAck(msgId, properties, null);
+                    return doCumulativeAck(msgId, properties, null, false);
                 } else if (batchIndexAckEnabled) {
-                    return doCumulativeBatchIndexAck(batchMessageId, properties);
+                    return doCumulativeBatchIndexAck(batchMessageId, properties, false);
                 } else {
-                    doCumulativeAck(MessageIdAdvUtils.prevMessageId(batchMessageId), properties, null);
+                    doCumulativeAck(MessageIdAdvUtils.prevMessageId(batchMessageId), properties, null, false);
                     return CompletableFuture.completedFuture(null);
                 }
             default:
                 throw new IllegalStateException("Unknown AckType: " + ackType);
         }
     }
 
-    private CompletableFuture<Void> doIndividualAck(MessageIdAdv messageId, Map<String, Long> properties) {
-        if (acknowledgementGroupTimeMicros == 0 || (properties != null && !properties.isEmpty())) {
+    private CompletableFuture<Void> doIndividualAck(MessageIdAdv messageId, Map<String, Long> properties,
+                                                    boolean queueDueToConnecting) {
+        if (!queueDueToConnecting
+                && (acknowledgementGroupTimeMicros == 0 || (properties != null && !properties.isEmpty()))) {
```

Comment:
Will it break the behavior of 

```
// We cannot group acks if the delay is 0 or when there are properties attached to it. Fortunately that's an
// uncommon condition since it's only used for the compaction subscription.
```
If the consumer is reconnecting but the ack has properties. We will also group the acks which is not expected?

Reply:
@codelipenghui Yes. But this behavior is only added for the `TwoPhaseCompactor`, the client side has no way to attach properties. see https://github.com/apache/pulsar/blob/55520bd810d29991d4af2124151c8d75c25b0800/pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java#L229-L230

Currently, if `acknowledge` is called during reconnection, `phaseTwoSeekThenLoop` will fail.

From my perspective, we should also queue the ACK requests and flush them after connected. @poorbarcode

Reply:
Agress with @BewareMyPower 

I am trying to split this PR into the following.
- add a new component `AcknowledgmentCache` to support caching the acknowledgments which include the arg `properties.`
- split `PersistentAcknowledgmentsGroupingTracker` into two implementations:
  - cache and batch the acks
  - immediately ack

Reply:
Marked current PR as `Draft`

Reply:
> split PersistentAcknowledgmentsGroupingTracker into two implementations

It's reasonable. BTW, the design of the C++ client is better that:
- [AckGroupingEnabled](https://github.com/apache/pulsar-client-cpp/blob/main/lib/AckGroupingTrackerEnabled.cc): cache the acknowledgments and flush them due to the timeout or count
- [AckGrouingDisabled](https://github.com/apache/pulsar-client-cpp/blob/main/lib/AckGroupingTrackerDisabled.cc): acknowledge immediately

https://github.com/apache/pulsar-client-cpp/blob/72b7311aeef32e28a28e926da686aaf948e8f948/lib/ConsumerImpl.cc#L201C1-L215C6

Though the naming is not good (but consistent with other impl classes in the library)